plugin-server/src/main/ingestion-queues/batch-processing/each-batch-webhooks.ts

```diff
@@ -0,0 +1,180 @@
+    hookCannon: HookCommander,
+    statsd: StatsD | undefined,
+    concurrency: number
+): Promise<void> {
+    // similar to eachBatch function in each-batch.ts, but without the dependency on the KafkaJSIngestionConsumer
+    // & handling the different batching return type
+    const key = 'async_handlers_webhooks'
+    const batchStartTimer = new Date()
+    const loggingKey = `each_batch_${key}`
+    const { batch, resolveOffset, heartbeat, commitOffsetsIfNecessary, isRunning, isStale }: EachBatchPayload = payload
+
+    const transaction = Sentry.startTransaction({ name: `eachBatchWebhooks` })
+
+    try {
+        const batchesWithOffsets = groupIntoBatches(batch.messages, concurrency, actionMatcher)
+
+        statsd?.histogram('ingest_event_batching.input_length', batch.messages.length, { key: key })
+        statsd?.histogram('ingest_event_batching.batch_count', batchesWithOffsets.length, { key: key })
+
+        for (const { eventBatch, lastOffset, lastTimestamp } of batchesWithOffsets) {
+            const batchSpan = transaction.startChild({ op: 'messageBatch', data: { batchLength: eventBatch.length } })
+
+            if (!isRunning() || isStale()) {
+                status.info('🚪', `Bailing out of a batch of ${batch.messages.length} events (${loggingKey})`, {
+                    isRunning: isRunning(),
+                    isStale: isStale(),
+                    msFromBatchStart: new Date().valueOf() - batchStartTimer.valueOf(),
+                })
+                await heartbeat()
+                return
+            }
+
+            await Promise.all(
+                eventBatch.map((event: RawClickHouseEvent) =>
+                    eachMessageWebhooksHandlers(event, actionMatcher, hookCannon, statsd).finally(() => heartbeat())
+                )
+            )
+
+            resolveOffset(lastOffset)
+            await commitOffsetsIfNecessary()
+
+            // Record that latest messages timestamp, such that we can then, for
+            // instance, alert on if this value is too old.
+            latestOffsetTimestampGauge
+                .labels({ partition: batch.partition, topic: batch.topic, groupId: key })
+                .set(Number.parseInt(lastTimestamp))
+
+            await heartbeat()
+
+            batchSpan.finish()
+        }
+
+        status.debug(
+            '🧩',
+            `Kafka batch of ${batch.messages.length} events completed in ${
+                new Date().valueOf() - batchStartTimer.valueOf()
+            }ms (${loggingKey})`
+        )
+    } finally {
+        statsd?.timing(`kafka_queue.${loggingKey}`, batchStartTimer)
+        transaction.finish()
+    }
+}
+
+export async function eachMessageWebhooksHandlers(
+    clickHouseEvent: RawClickHouseEvent,
+    actionMatcher: ActionMatcher,
+    hookCannon: HookCommander,
+    statsd: StatsD | undefined
+): Promise<void> {
+    if (!actionMatcher.hasWebhooks(clickHouseEvent.team_id)) {
+        // exit early if no webhooks nor resthooks
```

Comment:
nit: not needed anymore, as we dropped the events earlier

Reply:
yup, but no reason not to keep in case something changes in the future